src/v/kafka/server/quota_manager.cc

```diff
@@ -501,101 +505,51 @@ quota_manager::get_global_map_for_testing() const {
 // erase inactive tracked quotas. windows are considered inactive if
 // they have not received any updates in ten window's worth of time.
 void quota_manager::gc() {
-    // TODO: reimplement gc()
-    // vassert(
-    //   ss::this_shard_id() == _client_quotas.shard_id(),
-    //   "gc should only be performed on the owner shard");
-    // auto full_window = _default_num_windows() * _default_window_width();
-    // auto expire_threshold = clock::now() - 10 * full_window;
-    // ssx::background
-    //   = ssx::spawn_with_gate_then(_gate, [this, expire_threshold]() {
-    //         return do_gc(expire_threshold);
-    //     }).handle_exception([](const std::exception_ptr& e) {
-    //         vlog(klog.warn, "Error garbage collecting quotas - {}", e);
-    //     });
+    auto full_window = _default_num_windows() * _default_window_width();
+    auto expire_threshold = clock::now() - 10 * full_window;
+    ssx::background
+      = ssx::spawn_with_gate_then(_gate, [this, expire_threshold]() {
+            return container()
+              .invoke_on_all([expire_threshold](quota_manager& qm) {
+                  return qm.do_local_gc(expire_threshold);
+              })
+              .then([this]() { return do_global_gc(); });
+        }).handle_exception([](const std::exception_ptr& e) {
+            vlog(klog.warn, "Error garbage collecting quotas - {}", e);
+        });
 }
 
-namespace {
-
-/// A simplified copy paste of `std::set_intersection`. Copied here because we
-/// rely on the fact that we're allowed to have first1 == result for in-place
-/// set intersection in `do_gc()` which the contract of `std::set_intersection`
-/// does not allow despite that its implementation does.
-/// Requires: the inputs to be sorted
-/// Guarantees: the output to be sorted
-template<typename It1, typename It2, typename ItR>
-ItR set_intersection(It1 first1, It1 last1, It2 first2, It2 last2, ItR result) {
-    while (first1 != last1 && first2 != last2) {
-        if (*first1 < *first2) {
-            ++first1;
-        } else if (*first2 < *first1) {
-            ++first2;
-        } else { // *first1 == *first2
-            *result = *first1;
-            ++first1;
-            ++first2;
+void quota_manager::do_global_gc() {
+    vassert(
+      ss::this_shard_id() == quotas_shard,
+      "do_global_gc() should only be called on the owner shard");
+
+    for (auto it = _global_map->begin(); it != _global_map->end();) {
```

Comment:
I've now added some `ss::maybe_yield()`'s into these loops (see https://github.com/redpanda-data/redpanda/pull/19941#discussion_r1650625590), so I think now it makes more sense to stick with the iterator-based loop rather than `std::erase_if`.